### PR TITLE
1231 implementierung der regeln für die öffnungsuhrzeit

### LINKED ISSUE #1231: Implementierung der Regeln für die Öffnungsuhrzeit
# Akzeptanzkriterien
- [ ] Die Öffnungsuhrzeit kann nicht in der Zukunft liegen
- [ ] Die Öffnungsuhrzeit muss größer oder gleich der Wert der Konfiguration für früheste Öffnungsuhrzeit
- [ ] Die Öffnungsuhrzeit muss kleiner oder gleich der Wert der Konfiguration für späteste Öffnungsuhrzeit
- [ ] Die Bedeutung der Konfigurationsparameter `FRUEHESTE_EROEFFNUNGSZEIT_UW` und `FRUEHESTE_SCHLIESSUNGSZEIT_UW` ist im Code aussagekräftog Dokumentiert
- [ ] Der Regelverstoß ist für den Nutzer erkennbar
- [ ] Der Nutzer sieht gegen welche Regel verstoßen wurde
- [ ] Ein Speichern ist bei Regelverstoß nicht möglich

# Umsetzungshinweise
- verwendet den Store aus #1137

# Begründung Referenzstory
- 3 statt [2](#912) weil es mehrere Regeln sind

### PR DESCRIPTION
**Pull request by vjohnslhm**
# Beschreibung:

die regeln zur öffnung der wahlhandlung wurden implementiert

### Frontend
- [x] [Naming Conventions][naming-conventions-link] beachtet
- [x] Unit-Tests gepflegt
- [x] Komponententests gepflegt
- [x] Texte auf Rechtschreibung und Grammatik geprüft

# Referenzen[^1]:

Verwandt mit Issue #1356

Closes #1231

> [^1]: _Nicht zutreffende Referenzen vor dem Speichern entfernen_

[naming-conventions-link]: https://it-at-m.github.io/Wahllokalsystem/technik/naming_conventions/

## Summary by CodeRabbit

* **New Features**
  * Enhanced validation for election opening time input, ensuring it is required, not in the future, and within allowed opening and closing time parameters.
  * Time input now supports both "HH:MM" and "HH:MM:SS" formats.

* **Bug Fixes**
  * Improved error messages and validation feedback for invalid time entries in the election opening form.

* **Tests**
  * Added and updated tests to cover new time validation rules and boundary conditions.
  * Introduced snapshot tests to verify correct UI behavior for invalid time inputs.

### CODE REVIEW

File: wls-gui-wahllokalsystem/src/stores/infomanagementStore.ts [hunk 1]
```diff
@@ -23,9 +23,27 @@ export const useInfomanagementStore = defineStore(storeID, () => {
     }
   }
 
+  /** FRUEHESTE_EROEFFNUNGSZEIT bezeichnet den frühesten Wert, zu dem die Wahlhandlung eröffnet werden kann. */
+  const FRUEHESTE_EROEFFNUNGSZEIT_UW_PARAM = computed(() => {
```
Comment (MrSebastian) [inline wls-gui-wahllokalsystem/src/stores/infomanagementStore.ts]: Bitte die Wahlbezirksart unspezifische Form verwenden. Der Service nimmt basierend auf der WahlbezirkArt dann den korrekten Wert.

Das gilt auch für die Schließungsuhrzeit.

File: wls-gui-wahllokalsystem/src/util/rules.ts [hunk 2]
```diff
@@ -21,3 +25,15 @@ export const MIN_NUMBER = (min: number) => (value: number) =>
   value >= min || `Eingabe darf nicht kleiner als ${min} sein.`;
 
 export const NO_NEGATIVE_INPUT = MIN_NUMBER(0);
+
+export const TIME_NOT_IN_FUTURE = (value: string) =>
+  getDateFromTimeString(value) <= new Date() ||
+  `Eingabe darf nicht in der Zukunft liegen.`;
+
+export const TIME_GREATER_THAN = (compareValue: string) => (value: string) =>
```
Comment (MrSebastian) [inline wls-gui-wahllokalsystem/src/util/rules.ts]: Bitte `GREATER_OR_EQUAL` als Namen verwenden

File: wls-gui-wahllokalsystem/src/util/rules.ts [hunk 3]
```diff
@@ -21,3 +25,15 @@ export const MIN_NUMBER = (min: number) => (value: number) =>
...
+  getDateFromTimeString(value) >= getDateFromTimeString(compareValue) ||
+  `Eingabe muss größer als ${compareValue} sein.`;
```
Comment (MrSebastian) [inline wls-gui-wahllokalsystem/src/util/rules.ts]:
```suggestion
  `Eingabe muss größer oder gleich ${compareValue} sein.`;
```

File: wls-gui-wahllokalsystem/src/util/rules.ts [hunk 4]
```diff
@@ -21,3 +25,15 @@ export const MIN_NUMBER = (min: number) => (value: number) =>
...
+
+export const TIME_LESS_THAN = (compareValue: string) => (value: string) =>
```
Comment (MrSebastian) [inline wls-gui-wahllokalsystem/src/util/rules.ts]: Bitte `LESS_OR_EQUAL` verwenden

File: wls-gui-wahllokalsystem/src/util/rules.ts [hunk 5]
```diff
@@ -21,3 +25,15 @@ export const MIN_NUMBER = (min: number) => (value: number) =>
...
+  getDateFromTimeString(value) <= getDateFromTimeString(compareValue) ||
+  `Eingabe muss kleiner als ${compareValue} sein.`;
```
Comment (MrSebastian) [inline wls-gui-wahllokalsystem/src/util/rules.ts]:
```suggestion
  `Eingabe muss kleiner oder gleich ${compareValue} sein.`;
```

File: wls-gui-wahllokalsystem/src/util/rules.ts [hunk 6]
```diff
@@ -21,3 +25,15 @@ export const MIN_NUMBER = (min: number) => (value: number) =>
   value >= min || `Eingabe darf nicht kleiner als ${min} sein.`;
 
 export const NO_NEGATIVE_INPUT = MIN_NUMBER(0);
+
```
Comment (MrSebastian) [inline wls-gui-wahllokalsystem/src/util/rules.ts:28]: zu den neuen Rules bitte Tests ergänzen

File: wls-gui-wahllokalsystem/src/stores/infomanagementStore.ts [hunk 7]
```diff
@@ -23,9 +23,27 @@ export const useInfomanagementStore = defineStore(storeID, () => {
...
+      (param) => param.schluessel === "FRUEHESTE_SCHLIESSUNGSZEIT_UW"
+    );
+    return param?.wert || "00:00:00";
+  });
+
   return {
     konfigurationsparameter,
     initKonfigurationsparameter,
+    FRUEHESTE_EROEFFNUNGSZEIT_UW_PARAM,
+    FRUEHESTE_SCHLIESSUNGSZEIT_UW_PARAM,
```
Comment (MrSebastian) [inline wls-gui-wahllokalsystem/src/stores/infomanagementStore.ts]: In meiner Implementierung (PR #1461) habe ich sie camelCase gelassen. Die Variante würde ich auch bevorzugen dass es nicht wirklich Konstanten sind, sondern reaktive Objekte.

File: wls-gui-wahllokalsystem/src/stores/infomanagementStore.ts [hunk 8]
```diff
@@ -23,9 +23,27 @@ export const useInfomanagementStore = defineStore(storeID, () => {
...
+  /** FRUEHESTE_EROEFFNUNGSZEIT bezeichnet den frühesten Wert, zu dem die Wahlhandlung eröffnet werden kann. */
+  const FRUEHESTE_EROEFFNUNGSZEIT_UW_PARAM = computed(() => {
+    const param = konfigurationsparameter.value?.find(
+      (param) => param.schluessel === "FRUEHESTE_EROEFFNUNGSZEIT_UW"
+    );
+    return param?.wert || "00:00:00";
+  });
+
+  /** FRUEHESTE_SCHLIESSUNGSZEIT bezeichnet den spätesten Wert, zu dem die Wahlhandlung eröffnet werden kann. */
+  const FRUEHESTE_SCHLIESSUNGSZEIT_UW_PARAM = computed(() => {
```
Comment (MrSebastian) [inline wls-gui-wahllokalsystem/src/stores/infomanagementStore.ts]: Der andere Wert ist doch die `SPAETESTE_EROEFFNUNGSZEIT`

**Review by MrSebastian — CHANGES_REQUESTED** [review]
Ich überlege ob es nicht bessere wäre wenn die Getter im Infomanagement-Store ein Date wären. Der Vorteil wäre dass nicht bei jedem Regelaufruf ein Datum erzeugt werden müsste. Allerdings müsste im Fehlerfall dann aus dem Datum ein Text gemacht werden.

Die Standardwerte sollten wir auch noch diskutieren. Ich bin mir nicht sicher ob der Standard besser ist, oder dass die Anwendung nicht startet oder dass beim Fehlen bzw. inkorrekten Daten es sich verhält als würde es keine Grenzen geben.

File: wls-gui-wahllokalsystem/src/stores/infomanagementStore.ts (hunk 8 above)
Comment (vjohnslhm) [reply wls-gui-wahllokalsystem/src/stores/infomanagementStore.ts]: nope. in den issues #1323 und #1356 steht es nochmal drin, die späteste eröffnungsuhrzeit ist nur der trigger für den dialog, aber nicht der grenzwert für den user input

File: wls-gui-wahllokalsystem/src/stores/infomanagementStore.ts (hunk 7 above)
Comment (vjohnslhm) [reply wls-gui-wahllokalsystem/src/stores/infomanagementStore.ts]: [6696048](https://github.com/it-at-m/Wahllokalsystem/pull/1468/commits/66960486d6079287ce5d00239fc3b920e7ddb2e1)

File: wls-gui-wahllokalsystem/src/util/rules.ts (hunk 6 above)
Comment (vjohnslhm) [reply wls-gui-wahllokalsystem/src/util/rules.ts:28]: [d3814af](https://github.com/it-at-m/Wahllokalsystem/pull/1468/commits/d3814af0cbe8a4ab2e682e6892ee821c74c8fbf1)

File: wls-gui-wahllokalsystem/src/util/rules.ts (hunk 5 above)
Comment (vjohnslhm) [reply wls-gui-wahllokalsystem/src/util/rules.ts]: [39bdb27](https://github.com/it-at-m/Wahllokalsystem/pull/1468/commits/39bdb271af9b2b2acfb8d8d3e9e04aa9adb925ef)

File: wls-gui-wahllokalsystem/src/util/rules.ts (hunk 4 above)
Comment (vjohnslhm) [reply wls-gui-wahllokalsystem/src/util/rules.ts]: [39bdb27](https://github.com/it-at-m/Wahllokalsystem/pull/1468/commits/39bdb271af9b2b2acfb8d8d3e9e04aa9adb925ef)

File: wls-gui-wahllokalsystem/src/util/rules.ts (hunk 3 above)
Comment (vjohnslhm) [reply wls-gui-wahllokalsystem/src/util/rules.ts]: [39bdb27](https://github.com/it-at-m/Wahllokalsystem/pull/1468/commits/39bdb271af9b2b2acfb8d8d3e9e04aa9adb925ef)

File: wls-gui-wahllokalsystem/src/util/rules.ts (hunk 2 above)
Comment (vjohnslhm) [reply wls-gui-wahllokalsystem/src/util/rules.ts]: [39bdb27](https://github.com/it-at-m/Wahllokalsystem/pull/1468/commits/39bdb271af9b2b2acfb8d8d3e9e04aa9adb925ef)

**Comment by vjohnslhm** [comment]
> Ich überlege ob es nicht bessere wäre wenn die Getter im Infomanagement-Store ein Date wären. Der Vorteil wäre dass nicht bei jedem Regelaufruf ein Datum erzeugt werden müsste. Allerdings müsste im Fehlerfall dann aus dem Datum ein Text gemacht werden.

Da es im Endeffekt so oder so gempaaed werden muss - entweder beim fehler zu nem string, oder so wie jetzt zu nem date - habe ich es so gelassen wie es ist.

> Die Standardwerte sollten wir auch noch diskutieren. Ich bin mir nicht sicher ob der Standard besser ist, oder dass die Anwendung nicht startet oder dass beim Fehlen bzw. inkorrekten Daten es sich verhält als würde es keine Grenzen geben.

soll ich dazu ein issue erstellen / stefan direkt anschreiben? oder nehmen wir das mit in die demo?

**Comment by MrSebastian** [comment]
> > Die Standardwerte sollten wir auch noch diskutieren. Ich bin mir nicht sicher ob der Standard besser ist, oder dass die Anwendung nicht startet oder dass beim Fehlen bzw. inkorrekten Daten es sich verhält als würde es keine Grenzen geben.
> 
> soll ich dazu ein issue erstellen / stefan direkt anschreiben? oder nehmen wir das mit in die demo?

Ich habe es als Frage für die Demo zu Sprint 26 aufgenommen

File: wls-gui-wahllokalsystem/tests/unit/rules.spec.ts [hunk 9]
```diff
@@ -84,4 +87,70 @@ describe("Validation rules", () => {
...
+      });
+    });
+
+    afterEach(() => {
+      vi.useRealTimers();
+    });
+
+    const rule = TIME_NOT_IN_FUTURE;
+
+    it.each(["12:23", "15:00:00"])(
```
Comment (MrSebastian) [inline wls-gui-wahllokalsystem/tests/unit/rules.spec.ts]: Die Werte sollten einmal der gleiche wie von Mocked now und etwas davor zeigen um die Grenzwerte abzutesten.

File: wls-gui-wahllokalsystem/tests/unit/rules.spec.ts [hunk 10]
```diff
@@ -84,4 +87,70 @@ describe("Validation rules", () => {
...
+      "should_returnTrue_when_inputTime'%s'IsLessOrEqualToNow",
+      (input) => {
+        expect(rule(input)).toStrictEqual(true);
+      }
+    );
+
+    it.each(["30:25", "14:75", "10:20:98", "15:01", "20:00:00"])(
```
Comment (MrSebastian) [inline wls-gui-wahllokalsystem/tests/unit/rules.spec.ts]: da die Rule alle Arten von Strings annimmt bitte auch nicht Uhrzeiten als Parameter mit aufnehmen

File: wls-gui-wahllokalsystem/tests/unit/rules.spec.ts [hunk 11]
```diff
@@ -84,4 +87,70 @@ describe("Validation rules", () => {
...
+    });
+
+    afterEach(() => {
+      vi.useRealTimers();
+    });
+
+    const rule = TIME_NOT_IN_FUTURE;
+
+    it.each(["12:23", "15:00:00"])(
+      "should_returnTrue_when_inputTime'%s'IsLessOrEqualToNow",
```
Comment (MrSebastian) [inline wls-gui-wahllokalsystem/tests/unit/rules.spec.ts:107]: Für den False-False, d.h. es ist eine valide Uhrzeit, aber in der Zukunft, bitte einen expliziten Test analog zu diesem True-Cases erstellen. Ich denke der Fall geht bei `isInvalid` unter

File: wls-gui-wahllokalsystem/tests/unit/rules.spec.ts [hunk 12]
```diff
@@ -84,4 +87,70 @@ describe("Validation rules", () => {
...
+    const rule = TIME_LESS_OR_EQUAL("16:00");
+
+    it.each(["08:00", "12:36", "16:00:00"])(
+      "should_returnTrue_when_inputTime'%s'IsLessOrEqualToComparedValue",
+      (input) => {
+        expect(rule(input)).toStrictEqual(true);
+      }
+    );
+
+    it.each(["16:01", "26:14"])(
```
Comment (MrSebastian) [inline wls-gui-wahllokalsystem/tests/unit/rules.spec.ts]: den invalid-Fall mit 26Uhr würde ich separat testen, neben Tests zu beliebigen Strings.

File: wls-gui-wahllokalsystem/tests/unit/rules.spec.ts [hunk 13]
```diff
@@ -84,4 +87,70 @@ describe("Validation rules", () => {
...
+    const rule = TIME_GREATER_OR_EQUAL("08:00");
+
+    it.each(["08:00", "23:59:59", "15:00:00"])(
+      "should_returnTrue_when_inputTime'%s'IsGreaterOrEqualToComparedValue",
+      (input) => {
+        expect(rule(input)).toStrictEqual(true);
+      }
+    );
+
+    it.each(["00:00", "07:59"])(
```
Comment (MrSebastian) [inline wls-gui-wahllokalsystem/tests/unit/rules.spec.ts:138]: Bitte Tests ergänzen zu dem Fall das der zu prüfenden Werte keine valide Uhrzeit ist, sowie sonstige Strings

**Review by MrSebastian — CHANGES_REQUESTED** [review]

File: wls-gui-wahllokalsystem/src/stores/infomanagementStore.ts (hunk 8 above)
Comment (MrSebastian) [reply wls-gui-wahllokalsystem/src/stores/infomanagementStore.ts]: top, jetzt bin ich verwirrt .. gut das wir es dokumentiert haben. Ich glaube ich zeichne eine Grafik für die Doku :D

Comment (vjohnslhm) [reply wls-gui-wahllokalsystem/src/stores/infomanagementStore.ts]: stefan meinte wir können uns auch gerne bessere namen für die konstanten einfallen lassen, wenn das hilft. nur haben die das wohl auch schon versucht gehabt und keinen gefunden. das in die doku aufzunehmen finde ich eine gute idee 👍🏽 dafür würde ich aber dann #1356 wieder öffnen dann ist der kontext da oder?

File: wls-gui-wahllokalsystem/tests/unit/rules.spec.ts (hunk 13 above)
Comment (vjohnslhm) [reply wls-gui-wahllokalsystem/tests/unit/rules.spec.ts:138]: [706ba32](https://github.com/it-at-m/Wahllokalsystem/pull/1468/commits/706ba329b3c9e6bd1d37419f0ec5467f30277ae1)

File: wls-gui-wahllokalsystem/tests/unit/rules.spec.ts (hunk 12 above)
Comment (vjohnslhm) [reply wls-gui-wahllokalsystem/tests/unit/rules.spec.ts]: [706ba32](https://github.com/it-at-m/Wahllokalsystem/pull/1468/commits/706ba329b3c9e6bd1d37419f0ec5467f30277ae1)

File: wls-gui-wahllokalsystem/tests/unit/rules.spec.ts (hunk 11 above)
Comment (vjohnslhm) [reply wls-gui-wahllokalsystem/tests/unit/rules.spec.ts:107]: [706ba32](https://github.com/it-at-m/Wahllokalsystem/pull/1468/commits/706ba329b3c9e6bd1d37419f0ec5467f30277ae1)

File: wls-gui-wahllokalsystem/tests/unit/rules.spec.ts (hunk 10 above)
Comment (vjohnslhm) [reply wls-gui-wahllokalsystem/tests/unit/rules.spec.ts]: [706ba32](https://github.com/it-at-m/Wahllokalsystem/pull/1468/commits/706ba329b3c9e6bd1d37419f0ec5467f30277ae1)

File: wls-gui-wahllokalsystem/tests/unit/rules.spec.ts (hunk 9 above)
Comment (vjohnslhm) [reply wls-gui-wahllokalsystem/tests/unit/rules.spec.ts]: [706ba32](https://github.com/it-at-m/Wahllokalsystem/pull/1468/commits/706ba329b3c9e6bd1d37419f0ec5467f30277ae1)

**Review by MrSebastian — APPROVED** [review]
